application/src/main/java/org/thingsboard/server/service/ttl/KafkaEdgeTopicsCleanUpService.java

```diff
@@ -0,0 +1,107 @@
+ * limitations under the License.
+ */
+package org.thingsboard.server.service.ttl;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+import org.thingsboard.server.common.data.AttributeScope;
+import org.thingsboard.server.common.data.id.EdgeId;
+import org.thingsboard.server.common.data.id.TenantId;
+import org.thingsboard.server.common.data.kv.AttributeKvEntry;
+import org.thingsboard.server.common.data.page.PageDataIterable;
+import org.thingsboard.server.common.msg.queue.ServiceType;
+import org.thingsboard.server.dao.attributes.AttributesService;
+import org.thingsboard.server.dao.edge.EdgeService;
+import org.thingsboard.server.dao.tenant.TenantService;
+import org.thingsboard.server.queue.discovery.PartitionService;
+import org.thingsboard.server.queue.discovery.TopicService;
+import org.thingsboard.server.queue.kafka.TbKafkaAdmin;
+import org.thingsboard.server.queue.kafka.TbKafkaSettings;
+import org.thingsboard.server.queue.kafka.TbKafkaTopicConfigs;
+import org.thingsboard.server.queue.util.TbCoreComponent;
+
+import java.time.Instant;
+import java.util.Date;
+import java.util.concurrent.TimeUnit;
+
+import static org.thingsboard.server.service.state.DefaultDeviceStateService.LAST_CONNECT_TIME;
+
+@Slf4j
+@Service
+@TbCoreComponent
+@RequiredArgsConstructor
+@ConditionalOnExpression("'${queue.type:null}'=='kafka' && ${edges.enabled:true} && ${sql.ttl.edge_events.edge_events_ttl:0} > 0")
+public class KafkaEdgeTopicsCleanUpService {
+
+    private final EdgeService edgeService;
+    private final TenantService tenantService;
+    private final AttributesService attributesService;
+
+    private final TopicService topicService;
+    private final PartitionService partitionService;
+
+    private final TbKafkaSettings kafkaSettings;
+    private final TbKafkaTopicConfigs kafkaTopicConfigs;
+
+    @Value("${sql.ttl.edge_events.edge_events_ttl:2628000}")
+    private long ttlSeconds;
+
+    @Scheduled(initialDelayString = "#{T(org.apache.commons.lang3.RandomUtils).nextLong(0, ${sql.ttl.edge_events.execution_interval_ms})}", fixedDelayString = "${sql.ttl.edge_events.execution_interval_ms}")
+    public void cleanUp() {
+        PageDataIterable<TenantId> tenants = new PageDataIterable<>(tenantService::findTenantsIds, 10_000);
+        for (TenantId tenantId : tenants) {
+            try {
+                cleanUp(tenantId);
+            } catch (Exception e) {
+                log.warn("Failed to drop kafka topics for tenant {}", tenantId, e);
+            }
+        }
+    }
+
+    private void cleanUp(TenantId tenantId) throws Exception {
+        if (!partitionService.resolve(ServiceType.TB_CORE, tenantId, tenantId).isMyPartition()) {
```

Comment:
Rewrite this to the following logic (to be 100% safe):

1. Fetch all kafka topics that match the `{prefix}tb_edge_event.notifications.*`. 
2. Then extract the tenant id and edge id. Group by tenant id.
3. Check that tenant id is valid (CE/PE - exists. PaaS - exists + active).
4. Check that edge is active (your logic)
5. Delete all topics that are invalid.

Reply:
Also, add a limit on number of edges into tenant profile. The initial limit may be copied from the same value of number of allowed devices (via upgrade script in SQL).

Reply:
Thanks for the comment! Fixed the first part! 
And I'll add a limit in a separate PR.